File: server/games/time_limited_modifier_base.py
from .modifier_base import ModifierBase


class TimeLimitedModifierBase(ModifierBase):
    def __init__(self):
        super().__init__()

        self.duration = 0           # in server ticks (30 ms / tick)
        self.ticks = 0              # ticks count since activation

    def on_update(self, game):
        if not self.active:
            return

        self.ticks -= 1
        if self.ticks <= 0:
            self.deactivate(game)

    def on_activation(self, game):
        """Called when the modifier is activated"""
        self.ticks = self.duration

File: app/services/games/timeLimitedModifierBase.ts
import type { GameBase } from "./gameBase";
import { ModifierBase, ModifierStatus } from "./modifierBase";

export abstract class TimeLimitedModifierBase extends ModifierBase {
  protected ticks: number = 0;
  protected duration: number = 0;

  constructor() {
    super();
  }

  activate(game: GameBase): void {
    super.activate(game);
    this.ticks = this.duration;
  }

  // Triggered events
  onUpdate(game: GameBase): void {
    if (this.status !== ModifierStatus.ACTIVE) return;

    this.ticks--;
    if (this.ticks <= 0) this.deactivate(game);
  }

  // Getters & Setters
  setDuration(duration: number): void {
    this.duration = duration;
  }

  getState(): Record<string, any> {
    return {
      ...super.getState(),
      ticks: this.ticks,
      duration: this.duration,
    };
  }
}
